raiden-ts/src/transport/epics/messages.ts

```diff
@@ -54,129 +52,182 @@ function getMessageBody(message: string | Signed<Message>): string {
+  }
+
   let retries = -1;
-  return defer(() => {
-    if (!start) start = Date.now();
-    return via$;
-  }).pipe(
-    switchMap((via) => {
-      if (!via) return EMPTY;
+  return deps.matrix$.pipe(
+    mergeMap(async (matrix) => {
       retries++;
-      if (typeof via !== 'string')
-        return defer(async () => {
-          via.send(content.body);
-          return via.label; // via RTC channel, complete immediately
-        });
-      else
-        return matrix$.pipe(
-          mergeMap(async (matrix) => matrix.sendToDevice(type, { [via]: { '*': content } })),
-          mapTo(via), // via toDevice message
-          // complete only when peer is online
-          delayWhen(
-            constant(
-              presence$.pipe(
-                filter(
-                  (presence) => presence?.payload.userId === via && presence.payload.available,
-                ),
-              ),
-            ),
-          ),
-        );
+      return matrix.sendToDevice('m.room.message', payload);
     }),
-    retryWhile(intervalFromConfig(config$), {
+    retryWhile(intervalFromConfig(deps.config$), {
       maxRetries: 3,
       onErrors: networkErrors,
-      log: log.warn,
+      log: deps.log.warn,
+    }),
+    // a higher-order observable is used to be able to have requests serialized/batched,
+    // but return observables for success/failures which will resolve independently when peer comes
+    // online, and get merged in parallel at the end to avoid holding concatMap
+    mergeMap(function* () {
+      const tookMs = Date.now() - start;
+      for (const [peerId, peerQueue] of batch) {
+        for (const action of peerQueue) {
+          // wait independently on each peer's presence when resolving to success
+          yield deps.latest$.pipe(
+            filter(({ presences }) => presences[action.meta.address].payload.available),
+            take(1),
+            mapTo(messageSend.success({ via: peerId, tookMs, retries }, action.meta)),
+          );
+        }
+      }
+    }),
+    catchError(function* (err) {
+      for (const [, peerQueue] of batch) {
+        for (const action of peerQueue) {
+          // if failure, fail all queued responses at once, no need to wait
+          yield of(messageSend.failure(err, action.meta));
+        }
+      }
+    }),
+  );
+}
+
+function sendBatchedToDevice(
+  action$: Observable<messageSend.request>,
+  deps: RaidenEpicDeps,
+): Observable<messageSend.success | messageSend.failure> {
+  const queue: messageSend.request[] = [];
+  return action$.pipe(
+    tap((action) => queue.push(action)),
+    // like concatBuffer, but takes queue control in order to do custom batch selection
+    concatMap(() =>
+      deps.latest$.pipe(
+        first(),
+        // completeWith with some delay gives a minimal time for queue to be flushed
+        mergeMap((latest) =>
+          flushQueueToDevice(queue, latest, deps).pipe(completeWith(action$, 10)),
```

Comment:
They are just left in the queue for the next iteration. Each queued message will schedule a `concatMap` call which should return an observable; by the time its callback gets called, if the queue is empty (because all messages already got flushed), it'll `return EMPTY`, which is noop; otherwise, any leftover message will be picked up, with the first-come-first-served `msgtype` being chosen and dictating the msgtypes which will be popped for this peer.
In practice, the filtered messages which stay in queue should be picked up **as soon as** the previous `sendToDevice` call completes, so the only waiting time for them is the previous HTTP request to complete. The `success` (or failure) is sent only for the selected messages, i.e. the ones which got popped from the queue.

Reply:
Okay, I think I got that. Thanks for the explanation. :pray: